server/src/main/java/org/apache/iotdb/db/metadata/mtree/store/disk/schemafile/SchemaFile.java

```diff
@@ -277,157 +267,213 @@ public void writeMNode(IMNode node) throws MetadataException, IOException {
-          curPage.setNextSegAddress(curSegIdx, newSegAddr);
-
-          curSegAddr = newSegAddr;
+        if (curPage.getAsSegmentedPage().getSegmentSize(getSegIndex(actualAddress))
+            == SchemaPage.SEG_MAX_SIZ) {
+          // cur is a leaf, split and set next ptr as link between
+          ISchemaPage newPage = getMinApplSegmentedPageInMem(SchemaPage.SEG_MAX_SIZ);
+          newPage.getAsSegmentedPage().allocNewSegment(SchemaPage.SEG_MAX_SIZ);
+          sk =
+              curPage
+                  .getAsSegmentedPage()
+                  .splitWrappedSegment(
+                      entry.getKey(), childBuffer, newPage, SchemaPage.INCLINED_SPLIT);
+          curPage
+              .getAsSegmentedPage()
+              .setNextSegAddress((short) 0, getGlobalIndex(newPage.getPageIndex(), (short) 0));
+          if (treeTrace.get()[0] < 1) {
+            // first leaf to split and transplant, so that curSegAddr stay unchanged
+            ISchemaPage trsPage = getMinApplSegmentedPageInMem(SchemaPage.SEG_MAX_SIZ);
+            trsPage
+                .getAsSegmentedPage()
+                .transplantSegment(curPage.getAsSegmentedPage(), (short) 0, SchemaPage.SEG_MAX_SIZ);
+            ISchemaPage repPage =
+                ISchemaPage.initInternalPage(
+                    ByteBuffer.allocate(SchemaPage.PAGE_LENGTH),
+                    curPage.getPageIndex(),
+                    trsPage.getPageIndex());
+
+            repPage.getAsInternalPage().insertRecord(sk, newPage.getPageIndex());
+            repPage
+                .getAsInternalPage()
+                .setNextSegAddress(getGlobalIndex(trsPage.getPageIndex(), (short) 0));
+            replacePageInCache(repPage);
+
+            dirtyPages.put(trsPage.getPageIndex(), trsPage);
+          } else {
+            insertIndexEntry(treeTrace.get()[0], sk, newPage.getPageIndex());
+            dirtyPages.put(curPage.getPageIndex(), curPage);
+          }
         } else {
-          // segment on a single new page
-          curSegAddr = newPage.transplantSegment(curPage, curSegIdx, newSegSize);
-
-          curPage.deleteSegment(curSegIdx);
-
-          curSegIdx = SchemaFile.getSegIndex(curSegAddr);
+          // transplant and delete former segment
+          short actSegId = getSegIndex(actualAddress);
+          short newSegSize =
+              reEstimateSegSize(
+                  curPage.getAsSegmentedPage().getSegmentSize(actSegId) + childBuffer.capacity());
+          ISchemaPage newPage = getMinApplSegmentedPageInMem(newSegSize);
+
+          // with single segment, curSegAddr equals actualAddress
+          curSegAddr =
+              newPage
+                  .getAsSegmentedPage()
+                  .transplantSegment(curPage.getAsSegmentedPage(), actSegId, newSegSize);
+          newPage.getAsSegmentedPage().write(getSegIndex(curSegAddr), entry.getKey(), childBuffer);
+          curPage.getAsSegmentedPage().deleteSegment(actSegId);
```

Comment:
Use a new variable to represent ```SegmentedPage``` rather than always convert.

Reply:
Intensive convert had been replaced with an extra interface ```ISegmentedPage``` which extends ```ISchemaPage```.